feat (backend): implement matching service

DanielJames0302 opened the pull request:
# PR Description: Backend Matching Service Setup

## Overview
This PR implements a matching service that enables users to find suitable practice partners based on topic and difficulty preferences. The implementation includes a new microservice, Kafka-based asynchronous communication, Redis-based queue management, and enhanced frontend UI.

## Key Features

### 1. Matching Service
- **TypeScript-based microservice** running on port 8002
- **Socket.IO integration** for real-time client communication
- **JWT-based authentication** with debug mode support
- **Redis queue management** with sorted sets for efficient matching
- **Kafka integration** for async communication with question and collaboration services
- **Background worker** that runs matching algorithm every 1 second (configurable)
- **Distributed locking** using Redis to prevent race conditions
- **Graceful shutdown** with cleanup of resources

### 2. Matching Algorithm
- Compatibility checking based on topic and difficulty preferences
- "All" wildcard support for flexible matching
- Smart topic/difficulty resolution when users have different preferences
- Automatic cleanup of disconnected users from queue
- Prevention of duplicate matches

### 3. Kafka Event-Driven Architecture
Four Kafka topics orchestrate the matching flow:

1. **match_topic**: Matching service publishes match events
2. **question_topic**: Question service consumes and responds with questionId
3. **room_creation_topic**: Matching service requests room creation
4. **room_created_topic**: Collaboration service publishes roomId back

### 4. Frontend Enhancements
- **Timer overlay** with countdown (60 seconds default)
- **Real-time status updates**: Searching → Room Preparing → Match Found
- **Global timer** persists across page navigation
- **Cancel matching** functionality
- **Enhanced UX** with visual feedback and toast notifications
- **Improved state management** using Zustand with socket listeners

### 5. Integration Updates

#### Collaboration Service
- Added Kafka consumer for room creation requests
- Publishes room created events back to Kafka
- Enhanced room management with Kafka events

#### Question Service
- Added Kafka consumer for match events
- Returns matching question based on topic and difficulty
- Publishes questionId to Kafka for matching service

### 6. Infrastructure Updates
- **Docker Compose**: Added Redis, Kafka, Zookeeper services
- **Matching service container** with multi-stage Dockerfile
- **Environment configuration** for all services
- **CI/CD**: Added GitHub Actions workflow for matching service

## Technical Implementation

### Architecture Flow
1. User selects difficulty/topic on frontend
2. Frontend emits `matchStart` event via Socket.IO
3. Matching service adds user to Redis sorted queue
4. Background worker (1s interval) finds compatible matches
5. When match found:
   - Emit `roomPreparing` event
   - Publish match event to Kafka (match_topic)
6. Question service consumes event, fetches question, publishes questionId
7. Matching service requests room creation via Kafka
8. Collaboration service creates room, publishes roomId
9. Matching service emits `matchSuccess` with roomId
10. Frontend redirects to practice room

### Key Components

**Matching Service**:
- `matchingController.ts`: Handles socket events and countdown timers
- `matchingWorker.ts`: Background matching algorithm
- `kafka.ts`: Kafka manager with retry logic
- `redis.ts`: Redis client with data management
- `auth-socket.ts`: JWT authentication middleware

**Frontend**:
- `matching-store.ts`: Zustand state management
- `match-make.tsx`: Main matching interface
- `timer-overlay.tsx`: Persistent countdown component

## Configuration

### Environment Variables (Matching Service)
```env
PORT=8002
NODE_ENV=production
REDIS_URL=redis://redis:6379
KAFKA_HOST=kafka
KAFKA_PORT=9092
KAFKA_BROKERS=kafka:9092
QUESTION_SERVICE_URL=http://question-service:8003
COLLABORATION_SERVICE_URL=http://collaboration-service:8004
JWT_SECRET=<your-secret>
MATCH_COUNTDOWN_SECONDS=60
DEBUG_MODE=false
```

## Files Changed

### New Files (Matching Service)
- `backend/matching-service/**` (entire new service)
- `.github/workflows/ci-matching-service.yml`
- `backend/collaboration-service/src/config/kafka.js`
- `backend/collaboration-service/src/consumers/roomCreationConsumer.js`
- `backend/question-service/src/config/kafka.js`

### Modified Files
- `docker-compose.yml`: Added Redis, Kafka, Zookeeper, matching-service
- `frontend/src/app/matching/page.tsx`: Enhanced routing logic
- `frontend/src/stores/matching-store.ts`: Improved state management
- `frontend/src/components/matching/match-make.tsx`: Better UI/UX
- `frontend/src/components/matching/timer-overlay.tsx`: Timer improvements
- `backend/collaboration-service/src/controllers/roomController.js`: Kafka integration
- `backend/question-service/src/controllers/questionController.js`: Kafka consumer

## Testing Strategy

See detailed testing instructions below.

## Dependencies
- Added: `kafkajs`, `redis`, `socket.io`, `jsonwebtoken`, `uuid`
- Updated: Various service packages for Kafka integration

## Deployment Notes
- Requires Redis, Kafka, and Zookeeper to be running
- Ensure all environment variables are set
- Docker Compose will handle service dependencies
- Matching service depends on user-service, question-service, and collaboration-service

# Testing Instructions for Matching Service

## Prerequisites

1. **Environment Setup**
   - Docker and Docker Compose installed
   - All environment variables configured
   - Services: User Service, Question Service, Collaboration Service

2. **Start All Services**
   ```bash
   docker-compose up --build
   ```

   This will start:
   - Redis (port 6379)
   - MongoDB (port 27017)
   - Kafka & Zookeeper (ports 9092, 29092, 2181)
   - User Service (port 8001)
   - Question Service (port 8003)
   - Collaboration Service (ports 8004, 8005)
   - Matching Service (port 8002)
   - Frontend (port 3000)

## Manual Testing Steps

### Test 1: Single User Basic Flow

**Objective**: Verify timer and basic matching UI

1. Open browser and navigate to `http://localhost:3000`
2. Log in or create an account
3. Navigate to `/matching`
4. Select a difficulty (e.g., Easy) and topic (e.g., Arrays)
5. Click "Find your practice partner"
6. **Expected**: 
   - Timer overlay appears showing countdown from 60
   - Status shows "Finding your practice partner..."
   - Timer is visible and persists on the page

**Validate**: 
- ✅ Timer counts down correctly
- ✅ Timer is visible (fixed position at top)
- ✅ Can navigate to other pages while timer runs

### Test 2: Two Users Matching (Same Preferences)

**Objective**: Verify successful matching between two compatible users

1. **User A**: Open first browser/incognito window
   - Log in with account A
   - Go to `/matching`
   - Select "Medium" difficulty and "Arrays" topic
   - Click "Find your practice partner"

2. **User B**: Open second browser/incognito window
   - Log in with different account B
   - Go to `/matching`
   - Select "Medium" difficulty and "Arrays" topic
   - Click "Find your practice partner"

3. **Expected**:
   - Both users see countdown timer
   - Within 1-2 seconds, both see "Preparing your room..." message
   - Both users redirected to `/practice/{roomId}`
   - Same roomId for both users

**Validate**:
- ✅ Both users matched successfully
- ✅ Match found within 1-2 seconds
- ✅ Users redirected to practice room
- ✅ Check backend logs for matching flow

### Test 3: Two Users Matching (Compatible with "All")

**Objective**: Verify "all" wildcard matching

1. **User A**: Select "Easy" difficulty, leave topic empty (or select "all")
2. **User B**: Select "Easy" difficulty, select "Arrays" topic

**Expected**: 
- Users match successfully
- Final topic resolves to "Arrays" (from User B)

**Validate**:
- ✅ Matching works with wildcard preferences
- ✅ Question topic is correctly resolved

### Test 4: Matching Timeout

**Objective**: Verify timeout handling when no compatible match exists

**Setup**: Have only one user in the queue with specific preferences

1. Single user selects "Hard" difficulty and specific topic
2. No other users join or all have incompatible preferences
3. Wait for 60 seconds

**Expected**:
- Timer completes countdown
- Error message appears: "There are no active match requests that is suitable for your criteria..."
- User returned to matching page

**Validate**:
- ✅ Timeout message displayed correctly
- ✅ User removed from queue
- ✅ Can immediately start new match

### Test 5: Cancel Matching

**Objective**: Verify user can cancel matching request

1. Start matching with selected preferences
2. Wait a few seconds to see timer running
3. Click "Cancel" button on timer overlay

**Expected**:
- Timer disappears immediately
- Toast notification: "You've cancelled your matching request"
- User returned to matching selection page
- User removed from queue

**Validate**:
- ✅ Cancel works immediately
- ✅ Clean state reset
- ✅ Can start new match immediately after cancel

### Test 6: Multiple Incompatible Users

**Objective**: Verify users with incompatible preferences don't match

1. **User A**: Easy + Arrays
2. **User B**: Hard + Arrays (same topic, different difficulty)
3. **User C**: Medium + String (different topic, different difficulty)

**Expected**:
- Users A, B, C remain in queue
- No matches occur
- All see countdown timers running
- Eventually timeout after 60 seconds

**Validate**:
- ✅ Incompatible users don't match
- ✅ Queue maintains correct state
- ✅ Timeout occurs as expected

### Test 7: Page Navigation During Matching

**Objective**: Verify timer persists across page navigation

1. Start matching
2. While timer is running, navigate to Home page
3. Navigate to Account page
4. Navigate back to Matching page

**Expected**:
- Timer overlay stays visible on all pages
- Countdown continues uninterrupted
- Can cancel from any page

**Validate**:
- ✅ Timer persists across navigation
- ✅ State maintained correctly
- ✅ No socket reconnections (check network tab)

### Test 8: Multiple Matches (4+ Users)

**Objective**: Verify multiple pairs can be matched simultaneously

**Setup**: Have 4+ users ready in different browsers

1. **User A & User B**: Both select Easy + Arrays (should match)
2. **User C & User D**: Both select Medium + String (should match)
3. Start all matching requests simultaneously

**Expected**:
- A and B matched together
- C and D matched together
- All get different roomIds
- Matches occur within 1-2 seconds

**Validate**:
- ✅ Multiple concurrent matches work
- ✅ Each pair gets unique room
- ✅ No cross-contamination between matches

### Test 9: Kafka Event Flow Verification

**Objective**: Verify end-to-end Kafka event processing

1. Start a match between two users
2. Monitor backend logs:

**Expected Log Sequence**:
```
[Matching Service]
- Match found: socketId1 + socketId2
- Match event created: <matchId>
- Sent event to Kafka: match_topic
- Received question message: <questionId>
- Sent room creation request to Kafka
- Received room created message: <roomId>
- Sending matchSuccess event to users

[Question Service]
- Received match event from Kafka
- Selected question: <questionId>
- Published questionId to Kafka

[Collaboration Service]
- Received room creation request
- Created room: <roomId>
- Published roomId to Kafka
```

**Validate**:
- ✅ All Kafka events processed in correct order
- ✅ No duplicate processing
- ✅ Proper error handling at each stage

### Test 10: Disconnect Handling

**Objective**: Verify cleanup when user disconnects during matching

1. Start matching
2. Close browser tab/window abruptly
3. Try to match with another user

**Expected**:
- Disconnected user automatically removed from queue
- Second user can match with next available user
- No stale entries in Redis queue

**Validate**:
- ✅ Queue cleanup works correctly
- ✅ Can start new matches after disconnect
- ✅ Check Redis for any orphaned entries

### Test 11: Socket Reconnection

**Objective**: Verify socket reconnection handling

1. Start matching
2. Disconnect network briefly (disable WiFi for 2-3 seconds)
3. Reconnect network

**Expected**:
- Socket automatically reconnects
- Matching state restored
- Timer continues (though countdown may have advanced)
- Can still cancel or continue matching

**Validate**:
- ✅ Reconnection works smoothly
- ✅ State persistence
- ✅ No errors in console

### Test 12: Frontend State Management

**Objective**: Verify Zustand store maintains state correctly

1. Open browser DevTools
2. Check application state during matching:
   - `isMatching: true`
   - `count: <number>`
   - `selectedTopic: <topic>`
   - `connectionState: "connected"`

3. After match found:
   - `matchFound: true`
   - `roomId: <uuid>`
   - `questionId: <uuid>`

**Validate**:
- ✅ State transitions correctly
- ✅ No stale state
- ✅ Proper cleanup on unmount

## Debugging Tips

### Enable Debug Mode
Set `DEBUG_MODE=true` in matching service environment for easier testing without JWT.

### View Queue Status
```bash
# Access Redis CLI
docker exec -it <redis-container> redis-cli

# View matching queue
ZRANGE matching_queue 0 -1 WITHSCORES

# View user data
HGETALL user:<socketId>
```

### Monitor Kafka Topics
```bash
# List topics
docker exec kafka kafka-topics --list --bootstrap-server localhost:9092

# Consume messages
docker exec kafka kafka-console-consumer --topic match_topic --from-beginning
```

### View Service Logs
```bash
docker-compose logs -f matching-service
docker-compose logs -f question-service
docker-compose logs -f collaboration-service
```

## Common Issues and Solutions

### Issue: Users never match
**Solution**: Check Kafka topics are created, verify Redis connection, ensure all services are running

### Issue: Timer not showing
**Solution**: Check socket connection, verify JWT token, check browser console for errors

### Issue: Match found but no redirect
**Solution**: Verify roomId in state, check frontend routing, ensure practice page exists

### Issue: Kafka not connecting
**Solution**: Check Zookeeper is running, verify KAFKA_BROKERS env variable, check network connectivity

DanielJames0302 committed:
chore(pnpm-lock): update dependencies for backend matching service and add new packages

DanielJames0302 committed:
feat(docker-compose): add Redis and matching service to docker-compose configuration

- Introduced Redis service with persistent storage and network configuration.
- Added matching service with build context, environment variables, and dependencies on Redis.
- Updated volumes in docker-compose for Redis data storage.

chore(matching-service): enhance package.json and tsconfig.json

- Added build, start, and dev scripts to package.json for improved development workflow.
- Updated tsconfig.json for better TypeScript configuration, including target and module settings.

ncduy0303 commented on backend/matching-service/src/services/queueService.ts:
```diff
@@ -0,0 +1,147 @@
+    await redis.lPush(queueKey, JSON.stringify({
+      socketId,
+      roomId,
+      joinedAt: Date.now(),
```
1. Does `joinedAt` need to be the same value for  both the `userKey` and `queueKey`?
2. How is `joinedAt` being used?

DanielJames0302 committed:
Merge branch 'dev' into man/backend-matching-service-setup

DanielJames0302 committed:
feat(matching-service): integrate collaboration service and enhance matching logic

- Added support for collaboration service in the matching process, including the creation of collaboration rooms.
- Updated MatchRequest interface to allow optional difficulty and topic fields.
- Enhanced QueueService to manage user preferences more effectively, allowing for better matching based on difficulty and topic.
- Improved error handling and logging during collaboration room creation.
- Updated socket handlers to include collaboration room ID in match success responses.

DanielJames0302 committed:
refactor(queueService): remove joinedAt property and streamline queue management

DanielJames0302 committed:
feat(matching-service): enhance service with Kafka and Zookeeper integration

- Added Kafka and Zookeeper services to docker-compose for improved message handling.
- Updated matching service to include Kafka configuration and environment variables.
- Refactored matching service to remove deprecated methods and streamline code.
- Introduced new Redis logging functionality for queue status monitoring.
- Updated package dependencies and types for better compatibility and performance.

ho-thanh-und commented on backend/matching-service/src/models/types.ts:
```diff
@@ -0,0 +1,44 @@
+    MATH = 'MATH',
+    HASH_TABLES = 'HASH_TABLES',
+    LINKED_LISTS = 'LINKED_LISTS',
+  }
```
Since we are following the same chosen topics from the Project Docs, I suggest updating this to the following:

```TYPESCRIPT
export enum QUESTION_TOPIC {
     ARRAYS = 'ARRAYS',
     STRINGS = 'STRINGS',
     ALGORITHMS = 'ALGORITHMS',
     DATA_STRUCTURES = 'DATA_STRUCTURES',
     DATABASES = 'DATABASES',
     BIT_MANIPULATION = 'BIT_MANIPULATION',
     RECURSION = 'RECURSION',
     BRAINTEASER = 'BRAINTEASER',
}

```

DanielJames0302 committed:
Merge branch 'dev' into man/backend-matching-service-setup

DanielJames0302 committed:
feat(matching-service): enhance room creation and matching logic

- Updated room creation to accept optional question ID, user IDs, and programming language.
- Integrated Kafka message handling for question retrieval based on matching criteria.
- Improved error handling during collaboration room creation and added functionality to clear Redis data.
- Refactored matching logic to streamline user matching and added support for new socket events.
- Updated dependencies and configuration for better service integration.

DanielJames0302 committed:
Merge branch 'dev' into man/backend-matching-service-setup

DanielJames0302 committed:
feat(matching-service): enhance Kafka integration and room management

- Added Kafka configuration to docker-compose and environment variables for matching and collaboration services.
- Implemented new Kafka topics for room creation and handling room creation messages.
- Updated matching logic to process room creation requests and emit match success events to users.
- Refactored socket management to improve connection handling and added debug mode for socket authentication.
- Cleaned up unused code and improved error handling in matching and collaboration services.
- Removed deprecated components from the frontend and streamlined matching UI for better user experience.

DanielJames0302 committed:
fix(matching-page): update routing for match found to redirect to practice room

- Changed the redirection path from "/room" to "/practice/${roomId}" upon finding a match, improving the user flow for entering practice rooms.

ncduy0303 commented on docker-compose.yml:
```diff
@@ -43,12 +108,16 @@ services:
       - '8005:8005'
     environment:
-      - MONGODB_URI=mongodb://admin:password@mongodb:27017/peerprep?authSource=admin
+      - MONGODB_URI=mongodb://admin:password@mongodb:27017/peerprepCollabService?authSource=admin
```
```suggestion
      - MONGODB_URI=mongodb://admin:password@mongodb:27017/peerprepCollabServiceDB?authSource=admin
```

ncduy0303 commented on backend/collaboration-service/src/config.js:
```diff
@@ -3,7 +3,7 @@ export const config = {
   HTTP_PORT: process.env.HTTP_PORT || 8004,
   // Database
-  MONGO_URI: process.env.MONGODB_URI || "mongodb://localhost:27017/peerprep",
+  MONGO_URI: process.env.MONGODB_URI || "mongodb://admin:password@localhost:27017/peerprepCollabService?authSource=admin",
```
```suggestion
  MONGO_URI: process.env.MONGODB_URI || "mongodb://admin:password@localhost:27017/peerprepCollabServiceDB?authSource=admin",
```

ncduy0303 commented on backend/matching-service/src/config/kafka.ts:
```diff
@@ -0,0 +1,197 @@
+
+export const MATCH_TOPIC = 'match_topic';
+export const QUESTION_TOPIC = 'question_topic';
+export const ROOM_CREATION_TOPIC = 'room_creation_topic';
```
We might want to centralise all the Kafka topic names in one location outside all services later, but this is okay for now.

ncduy0303 commented on backend/matching-service/src/config/kafka.ts:
What is the reason for using 2 consumers instead of 1 consumer listening to 2 topics?

DanielJames0302 replied:
By having 2 consumers, this allows processing messages in 2 topics in parallel. 1 consumer is responsible for processing messages sent from Question Service while the other one will handle messages from Collaboration service

ncduy0303 commented on backend/matching-service/src/config/kafka.ts:
```diff
@@ -0,0 +1,197 @@
+    this.admin = this.kafka.admin();
+    this.producer = this.kafka.producer();
+    this.consumer = this.kafka.consumer({ groupId: 'matching-service-group' });
+    this.consumer2 = this.kafka.consumer({ groupId: 'matching-service-group-2' });
```
I think the naming for multiple consumers and groups can be clearer.

ncduy0303 commented on backend/question-service/src/controllers/questionController.js:
Do we still need GET /v1/questions/pick?topic=&difficulty= endpoint after this change?

DanielJames0302 replied:
I think it's still needed just in case the admin might need it

ncduy0303 commented on backend/question-service/src/controllers/questionController.js:
Does question service handle the case when topic or difficulty is set to `"all"` by the matching service (i.e., happened when both users don't set the topic or the difficulty)?

DanielJames0302 replied:
Yup ! I think that case is already handled in this code:
```
const pipeline = [];
const match = {};
if (topic) match.topic = topic;
if (difficulty) match.difficulty = difficulty;
if (Object.keys(match).length > 0) pipeline.push({ $match: match });
pipeline.push({ $sample: { size: 1 } });

const docs = await Question.aggregate(pipeline);
```

if either of them is not set, they won't be included in the `match`

ncduy0303 replied:
But in `matchingController.ts` and `matchingWorker.ts`, the matching service set the string `topic` and `difficulty` to "all" in the websocket message, so wouldn't the question service receive the string "all" instead of an empty string anymore?
```
socket.on(SOCKET_EVENTS.MATCH_START, async (requestData: { topic: string; difficulty: string }) => {
  let { topic, difficulty } = requestData
  
  if (topic === '') topic = 'all';
  if (difficulty === '') difficulty = 'all';
  ...
}
```

It might be good to define "all" as a constant for reuse too.

DanielJames0302 replied:
Fixed

ncduy0303 commented on backend/matching-service/src/workers/matchingWorker.ts:
```diff
@@ -0,0 +1,358 @@
+      matchId: matchId,
+      questionId: roomObject.questionId,
+      userIds: userIds,
+      programmingLanguage: 'python'
```
Better to not set any default programming language here, and leave the default set by the collab service itself.

DanielJames0302 replied:
I removed it in the code

ncduy0303 reviewed (changes requested):
Looks good overall. It would be nice to include some screenshots/recordings in the PR once done too.

DanielJames0302 committed:
Update docker-compose.yml

Co-authored-by: Nguyen Cao Duy <caoduy73@gmail.com>

DanielJames0302 committed:
refactor(docker-compose): add container names and dependencies for services

- Added container names for Redis, Zookeeper, and Kafka services for better identification.
- Updated Kafka service to depend on Zookeeper to ensure proper startup order.
- Cleaned up network definitions for clarity.
- Refactored Kafka consumer variable names for improved readability.

DanielJames0302 committed:
Merge branch 'dev' into man/backend-matching-service-setup

DanielJames0302 committed:
feat(kafka): define consumer group constants for improved readability

DanielJames0302 committed:
fix(matching-service): update matching logic to handle default values for topic and difficulty

- Set default values for topic and difficulty to 'all' in the matching process.
- Refactor matching logic to use constants for improved readability.
- Clean up console logs and comments in the question controller for clarity.
- Enable debug mode in docker-compose for easier troubleshooting.

DanielJames0302 committed:
fix(docker-compose): disable debug mode for production environment

DanielJames0302 committed:
chore(docker-compose): remove database configuration for MongoDB

ncduy0303 reviewed (approved):
lgmt, I think we can merge this after resolving conflicts. Can include some screenshot/recording in the PR description.
